_episodes/04-networks-are-like-onions.md

```diff
@@ -0,0 +1,529 @@
+
+
+~~~
+dim = train_images.shape[1] * train_images.shape[2] * train_images.shape[3]
+print(dim)
+~~~
+{: .language-python}
+~~~
+3072
+~~~
+{: .output}
+
+> ## Number of parameters
+>
+> Suppose we create a Dense (fully connected) layer with 100 hidden units as input units, how many parameters does this layer have?
+>
+> > ## Solution
+> >
+> > Each of the *dim* inputs is connected with 100 hidden units, and each of the hidden units has a bias. So we have *dim * 100 + 100* weights to learn.
+> > ~~~
+> > dim * 100 + 100
+> > ~~~
+> > {: .language-python}
+> > ~~~
+> > 307300
+> > ~~~
+> > {: .output}
+> > We can also check this by building the layer in Keras:
+> > ~~~
+> > inputs = keras.Input(shape=dim)
+> > outputs = keras.layers.Dense(100)(inputs)
+> > model = keras.models.Model(inputs=inputs, outputs=outputs)
+> > model.summary()
+> > ~~~
+> > {: .language-python}
+> > ~~~
+> >     Model: "model"
+    _________________________________________________________________
+    Layer (type)                 Output Shape              Param #
+    =================================================================
+    input_1 (InputLayer)         [(None, 3072)]            0
+    _________________________________________________________________
+    dense (Dense)                (None, 100)               307300
+    =================================================================
+    Total params: 307,300
+    Trainable params: 307,300
+    Non-trainable params: 0
+    _________________________________________________________________
+> > ~~~
+> > {: .output}
+> {: .solution}
+{: .challenge}
+
+
+
+Of course, we can decrease the number of units in our hidden layer, but this also decreases the number of patterns our network can remember. Moreover, if we increase the image size, this number of weights will explode, even though the task of recognizing large images is not necessarily more difficult than the task of recognizing small images.
+
+The solution is that we make the network learn in a 'smart' way. The features that we learn should be similar both for small and large images, and similar features (e.g. edges, corners) can appear anywhere in the image (in mathematical terms: *translation invariant*). We do this by making use of a concepts from image processing that precede Deep Learning.
+
+A **convolution matrix**, or **kernel**, is a matrix transformation that we 'slide' over the image to calculate features at each position of the image. For each pixel, we calculate the matrix product between the kernel and the pixel with its surroundings. A kernel is typically small, between 3x3 and 7x7 pixels. We can for example think of the simple 3x3 kernel:
+~~~
+[[0, 0, 0,],
+ [1, 1, 1],
+ [0, 0, 0]]
+~~~
+{: .language-python}
+This kernel will give a high value to a pixel if it's in the middle of a horizontal edge.
```

Comment:
This jumps directly to the conceptional consequence. This will introduce high mental load as for learners it may not be too obvious how the horizontal edge, i.e. high/low values of a fixed intensity related to the kernel. 

I suggest to align the text with this image:
https://github.com/dvgodoy/dl-visuals/blob/main/Convolutions/conv3.png

and then move onto a more explicit example like the one with horizontal edge.